File: go/compare.go
package main

import "fmt"

func main() {
    var constraintA = [][3]int{{0, 1, 0}, {1, 0, 1},  {1, 2, 0}}
    var constraintB = [][3]int{{0, 1, 1}, {1, 0, 1},  {1, 2, 0}}
    var constraintC = [][3]int{{2, 3, 0}, {2, 4, 1},  {3, 5, 0}, {1, 2, 0}}
    var constraintD = [][3]int{{2, 3, 1}, {2, 4, 0},  {3, 5, 1}, {1, 2, 1}}
    var rule1 = [][][3]int{constraintA, constraintB}
    var rule2 = [][][3]int{constraintC, constraintD}
    var all_rules = [][][][3]int{rule1, rule2}
    var families = []int{0, 1}
    var features = []int{1, 2}
    var loops int = 1000000
    var ans = [1000000]bool{}
    for i := 0; i < loops; i++ {
        ans[i] = check_unconstrained(all_rules, families, features)
    }
    fmt.Println(ans[1000000-1])
}

func check_unconstrained(all_rules [][][][3]int, families []int, features []int) bool {
    if len(all_rules) == 0 {
        return true
    } else if path_through_constraints(all_rules[0], families, features) {
        return check_unconstrained(all_rules[1:], families, features)
    } else {
        return false
    }
}

func path_through_constraints(constraints [][][3]int, families []int, features []int) bool {
    if len(constraints) == 0 {
        return false
    } else if constraint_check(constraints[0], families, features) {
        return true
    } else {
        return path_through_constraints(constraints[1:], families, features)
    }
}

func constraint_check(constraint [][3]int, families []int, features []int) bool {
    if len(constraint) == 0 {
        return true
    } else if check_this_feature(constraint[0], families, features) {
        return constraint_check(constraint[1:], families, features)
    } else {
        return false
    }
}

func check_this_feature(this_feature [3]int, families []int, features []int) bool {
    if have(this_feature) && family_included(this_feature, families) {
        return in(this_feature[1], features)
    } else if (!have(this_feature)) && family_included(this_feature, families) {
        return !in(this_feature[1], features)
    } else {
        return true
    }

}

func have(this_feature [3]int) bool {
    return this_feature[2] == 1
}

func family_included(this_feature [3]int, families []int) bool {
    return in(this_feature[0], families)
}

func in(val int, slice []int) bool {
    for item := range slice {
        if item == val {
            return true
        }
    }
    return false
}

// Run script: go run compare.go

File: rust/compare.rs
fn main() {
    let features: &[u16] = &[1, 2];
    let families: &[u16] = &[0, 1];
    let constraint_a: &[[u16; 3]] = &[[0, 1, 0], [1, 0, 1], [1, 2, 0]];
    let constraint_b: &[[u16; 3]] = &[[0, 1, 1], [1, 0, 1], [1, 2, 0]];
    let constraint_c: &[[u16; 3]] = &[[2, 3, 0], [2, 4, 1],  [3, 5, 0],  [1, 2, 0]];
    let constraint_d: &[[u16; 3]] = &[[2, 3, 1], [2, 4, 0],  [3, 5, 1],  [1, 2, 1]];
    let rule1: &[&[[u16; 3]]] = &[constraint_a, constraint_b];
    let rule2: &[&[[u16; 3]]] = &[constraint_c, constraint_d];
    let all_rules: &[&[&[[u16; 3]]]] = &[rule1, rule2];
    let loops: u32 = 1000000;
    let mut counter: u32 = 0;
    for _i in 0..loops {
        if check_unconstrained(all_rules, families, features) {
            counter = counter + 1;
        }
    }
    println!("{}", counter)

}

fn check_unconstrained(all_rules: &[&[&[[u16; 3]]]], families: &[u16], features: &[u16]) -> bool {
    return if all_rules.is_empty() {
        true
    } else if path_through_constraints(&all_rules[0], families, features) {
        check_unconstrained(&all_rules[1..], families, features)
    } else {
        false
    }
}


fn path_through_constraints(constraints: &[&[[u16; 3]]], families: &[u16], features: &[u16]) -> bool {
    return if constraints.is_empty() {
        false
    } else if constraint_check(&constraints[0], families, features) {
        true
    } else {
        path_through_constraints(&constraints[1..], families, features)
    }
}


fn constraint_check(constraint: &[[u16; 3]], families: &[u16], features: &[u16]) -> bool {
    return if constraint.is_empty() {
        true
    } else if check_this_feature(&constraint[0], families, features) {
        return constraint_check(&constraint[1..], families, features)
    } else {
        false
    };
}


fn check_this_feature(this_feature: &[u16; 3], families: &[u16], features: &[u16]) -> bool {
    let have_feature: &bool = &have(this_feature);
    let family_in: &bool = &family_included(this_feature, families);
    return if *have_feature && *family_in {
        features.contains(&this_feature[1])
    } else if !*have_feature && *family_in {
        !features.contains(&this_feature[1])
    } else {
        true
    };
}

fn have(this_feature: &[u16; 3]) -> bool {
    return this_feature[2] == 0;
}

fn family_included(this_feature: &[u16; 3], families: &[u16]) -> bool {
    return families.contains(&this_feature[0]);
}

// Compile command: rustc -o compare compare.rs;